aea/test_tools/test_cases.py

```diff
@@ -358,27 +383,148 @@ def _read_error(pid: subprocess.Popen):
+
+        cls.initialize_aea(cls.author)
+
+    @classmethod
+    def teardown_class(cls):
+        """Teardown the test."""
+        cls._terminate_subprocesses()
+        cls._join_threads()
+        cls.unset_agent_context()
+        cls.change_directory(cls.old_cwd)
+        cls.packages_dir_path = Path("packages")
+        cls.agents = set()
+        cls.current_agent_context = ""
+        try:
+            shutil.rmtree(cls.t)
+        except (OSError, IOError):
+            pass
+
+
+class UseOef:
+    @pytest.fixture(autouse=True)
+    def _start_oef_node(self, network_node):
+        """Start an oef node."""
+
+
+class AEATestCaseEmpty(BaseAEATestCase):
+    """
+    Test case for a default AEA project.
+
+    This test case will create a default AEA project.
+    """
+
+    @classmethod
+    def setup_class(cls):
+        """Set up the test class."""
+        BaseAEATestCase.setup_class()
+        cls.agent_name = "agent-" + "".join(random.choices(string.ascii_lowercase, k=5))
+        cls.create_agents(cls.agent_name)
+        cls.set_agent_context(cls.agent_name)
+
+
+class AEATestCaseMany(BaseAEATestCase):
+    """Test case for many AEA projects."""
+
+    @classmethod
+    def setup_class(cls):
+        """Set up the test class."""
+        BaseAEATestCase.setup_class()
+
+    @classmethod
+    def teardown_class(cls):
+        """Teardown the test."""
+        cls._terminate_subprocesses()
+        cls._join_threads()
+        try:
+            shutil.rmtree(cls.t)
+        except (OSError, IOError):
+            pass
+
+
+class AEATestCase(BaseAEATestCase):
+    """
+    Test case from an existing AEA project.
+
+    Subclass this class and set `path_to_aea` properly. By default,
+    it is assumed the project is inside the current working directory.
+    """
+
+    path_to_aea: Union[Path, str] = Path(".")
+    packages_dir_path = Path("..", "packages")
+    agent_configuration: AgentConfig
+
+    @classmethod
+    def setup_class(cls):
+        """Set up the test class."""
+        # make paths absolute
+        cls.path_to_aea = cls.path_to_aea.absolute()
+        cls.packages_dir_path = cls.packages_dir_path.absolute()
+        # load agent configuration
+        with Path(cls.path_to_aea, DEFAULT_AEA_CONFIG_FILE).open(
+            mode="r", encoding="utf-8"
+        ) as fp:
+            loader = ConfigLoader.from_configuration_type(PackageType.AGENT)
+            agent_configuration = loader.load(fp)
+        cls.agent_configuration = agent_configuration
+        cls.agent_name = agent_configuration.agent_name
+
+        # this will create a temporary directory and move into it
+        BaseAEATestCase.packages_dir_path = cls.packages_dir_path
+        BaseAEATestCase.setup_class()
+
+        # copy the content of the agent into the temporary directory
+        shutil.copytree(str(cls.path_to_aea), str(cls.t / cls.agent_name))
+        cls.set_agent_context(cls.agent_name)
```

Comment:
@marcofavorito why do you think it is necessary to create the temporary directory?

Reply:
because the test use case can potentially change the AEA project under testing, which is not desirable for tests.

Reply:
Fine!